compiler/rustc_type_ir/src/fast_reject.rs

```diff
@@ -214,184 +211,150 @@ impl<I: Interner> DeepRejectCtxt<I> {
-                TreatParams::ForLookup => false,
-                TreatParams::AsCandidateKey => true,
-            },
+            (ty::FnPtr(lhs_sig), ty::FnPtr(rhs_sig)) => {
+                let lhs_sig = lhs_sig.skip_binder();
+                let rhs_sig = rhs_sig.skip_binder();
 
-            ty::Infer(ty::IntVar(_)) => impl_ty.is_integral(),
-
-            ty::Infer(ty::FloatVar(_)) => impl_ty.is_floating_point(),
+                lhs_sig.abi == rhs_sig.abi
+                    && lhs_sig.c_variadic == rhs_sig.c_variadic
+                    && lhs_sig.safety == rhs_sig.safety
+                    && lhs_sig.inputs_and_output.len() == rhs_sig.inputs_and_output.len()
+                    && iter::zip(lhs_sig.inputs_and_output.iter(), rhs_sig.inputs_and_output.iter())
+                        .all(|(lhs, rhs)| self.types_may_unify(lhs, rhs))
+            }
+            (ty::FnDef(lhs_def_id, lhs_args), ty::FnDef(rhs_def_id, rhs_args))
+            | (ty::Closure(lhs_def_id, lhs_args), ty::Closure(rhs_def_id, rhs_args))
+            | (
+                ty::CoroutineClosure(lhs_def_id, lhs_args),
+                ty::CoroutineClosure(rhs_def_id, rhs_args),
+            )
+            | (ty::Coroutine(lhs_def_id, lhs_args), ty::Coroutine(rhs_def_id, rhs_args))
+            | (
+                ty::CoroutineWitness(lhs_def_id, lhs_args),
+                ty::CoroutineWitness(rhs_def_id, rhs_args),
+            ) => lhs_def_id == rhs_def_id && self.args_may_unify(lhs_args, rhs_args),
 
-            ty::Infer(_) => true,
+            (ty::Error(_), _) | (_, ty::Error(_)) => true,
 
             // As we're walking the whole type, it may encounter projections
             // inside of binders and what not, so we're just going to assume that
             // projections can unify with other stuff.
             //
             // Looking forward to lazy normalization this is the safer strategy anyways.
-            ty::Alias(..) => true,
+            (ty::Alias(..), _) | (_, ty::Alias(..)) => true,
 
-            ty::Error(_) => true,
+            // Bound type variables may unify with rigid types e.g. when using
+            // non-lifetime binders.
+            (ty::Bound(..), _) | (_, ty::Bound(..)) => true,
 
-            ty::CoroutineWitness(..) => {
-                panic!("unexpected obligation type: {:?}", obligation_ty)
-            }
+            (ty::Infer(var), _) => self.var_and_ty_may_unify(var, rhs),
+            (_, ty::Infer(var)) => self.var_and_ty_may_unify(var, lhs),
+
+            (ty::Param(p), _) => self.param_and_ty_may_unify(p, self.treat_lhs_params, rhs),
+            (_, ty::Param(p)) => self.param_and_ty_may_unify(p, self.treat_rhs_params, lhs),
+
+            // Placeholder types don't unify with anything on their own.
+            (ty::Placeholder(lhs), ty::Placeholder(rhs)) => lhs == rhs,
+            (ty::Placeholder(_), _) | (_, ty::Placeholder(_)) => false,
+
+            _ => false,
```

Comment:
this function has to be an overapproximation, so incorrectly returning `false` would result in subtle bugs. Please exhaustively match here: `(ty::Whatever | ty::Whatever | ...., _) | (_, ty::whatever | .....)`